server/data/data-library/src/test/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTableTest.java

```diff
@@ -267,91 +262,136 @@ public void testAliasDomainMapping() throws ErrorMappingException, RecipientRewr
+        MappingSource source = MappingSource.fromUser(USER, domain);
 
-        virtualUserTable.addMapping(source, Mapping.group(address));
+        virtualUserTable.addMapping(source, Mapping.group(ADDRESS));
         virtualUserTable.addMapping(source, Mapping.group(address2));
 
-        virtualUserTable.removeMapping(source, Mapping.group(address));
+        virtualUserTable.removeMapping(source, Mapping.group(ADDRESS));
         virtualUserTable.removeMapping(source, Mapping.group(address2));
 
-        assertThat(virtualUserTable.getMappings(user, domain))
+        assertThat(virtualUserTable.getMappings(USER, domain))
             .isEqualTo(MappingsImpl.empty());
     }
+
+    @Test
+    public void listSourcesShouldReturnWhenHasMapping() throws RecipientRewriteTableException {
+        Domain domain = Domain.LOCALHOST;
+        MappingSource source = MappingSource.fromUser(USER, domain);
+        Mapping mapping = Mapping.group(ADDRESS);
+        virtualUserTable.addMapping(source, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).contains(source);
+    }
+
+    @Test
+    public void listSourcesShouldReturnWhenMultipleSourceMapping() throws RecipientRewriteTableException {
+        Domain domain = Domain.of("james");
+
+        MappingSource source = MappingSource.fromUser(USER, domain);
+        MappingSource source2 = MappingSource.fromDomain(Domain.LOCALHOST);
+        Mapping mapping = Mapping.group(ADDRESS);
+
+        virtualUserTable.addMapping(source, mapping);
+        virtualUserTable.addMapping(source2, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).contains(source, source2);
+    }
+
+    @Test
+    public void listSourcesShouldReturnWhenHasForwardMapping() throws RecipientRewriteTableException {
+        Domain domain = Domain.LOCALHOST;
+        MappingSource source = MappingSource.fromUser(USER, domain);
+        Mapping mapping = Mapping.forward("forward");
+
+        virtualUserTable.addMapping(source, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).contains(source);
+    }
+
+    @Test
+    public void listSourcesShouldReturnWhenHasAddressMapping() throws RecipientRewriteTableException {
+        Domain domain = Domain.LOCALHOST;
+        MappingSource source = MappingSource.fromUser(USER, domain);
+        Mapping mapping = Mapping.address("address");
+
+        virtualUserTable.addMapping(source, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).contains(source);
+    }
+
+    @Test
+    public void listSourcesShouldReturnEmptyWhenHasRegexMapping() throws RecipientRewriteTableException {
+        Domain domain = Domain.LOCALHOST;
+        MappingSource source = MappingSource.fromUser(USER, domain);
+        Mapping mapping = Mapping.regex("regex");
+
+        virtualUserTable.addMapping(source, mapping);
+
+        assertThat(virtualUserTable.listSources(mapping)).isEmpty();
+    }
```

Comment:
What happens when we are listing sources for an Error mapping? a domain mapping? A regex mapping?

What happens when we are looking for a mapping that do not exist (and thus have no sources?)